File: dawn_go/ss.go
package main

import (
	"sort"
	"strings"
)

func get_ngrams(n int, s string) []string {
	var ngrams []string
	for i := 0; i <= len(s)-n; i++ {
		ngrams = append(ngrams, s[i:i+n])
	}
	return ngrams
}

func count_3gram_frequencies(sample_keys []string) map[string]int {
	frequencies := make(map[string]int)
	for _, key := range sample_keys {
		ngrams := get_ngrams(3, key)
		for _, ngram := range ngrams {
			frequencies[ngram]++
		}
	}
	return frequencies
}

func select_3gram_intervals(sample_keys []string, dict_size int) []struct {
	start string
	end   *string
} {
	frequencies := count_3gram_frequencies(sample_keys)
	sorted_frequencies := make([]struct {
		ngram string
		freq  int
	}, 0, len(frequencies))
	for ngram, freq := range frequencies {
		sorted_frequencies = append(sorted_frequencies, struct {
			ngram string
			freq  int
		}{ngram, freq})
	}
	sort.Slice(sorted_frequencies, func(i, j int) bool {
		return sorted_frequencies[i].freq > sorted_frequencies[j].freq
	})

	selected_ngrams := make([]string, 0, dict_size/2)
	for _, sf := range sorted_frequencies[:dict_size/2] {
		selected_ngrams = append(selected_ngrams, sf.ngram)
	}
	sort.Strings(selected_ngrams)

	var intervals []struct {
		start string
		end   *string
	}
	var last_ngram string
	for _, ngram := range selected_ngrams {
		start := last_ngram
		if len(start) > 0 {
			start = start[:len(start)-1] + string(rune(start[len(start)-1])+1)
		}
		intervals = append(intervals, struct {
			start string
			end   *string
		}{start, &ngram})
		last_ngram = ngram
	}
	return intervals
}

File: dawn_ocaml/lib/ss.ml
open Base

module SymbolSelector = struct
  let get_ngrams n s =
    let len = String.length s in
    let rec loop i acc =
      if i + n > len then List.rev acc
      else
        let ngram = String.sub s i n in
        loop (i + 1) (ngram :: acc)
    in
    loop 0 []

  let count_3gram_frequencies sample_keys =
    let trie = Hashtbl.create 1000 in
    List.iter
      (fun key ->
        let ngrams = get_ngrams 3 key in
        List.iter (fun ngram -> Hashtbl.replace trie ngram ((Hashtbl.find_opt trie ngram |> Option.value ~default:0) + 1)) ngrams)
      sample_keys;
    trie

  let select_3gram_intervals sample_keys dictionary_size =
    let frequencies = count_3gram_frequencies sample_keys in
    let sorted_frequencies = Hashtbl.fold (fun ngram freq acc -> (ngram, freq) :: acc) frequencies [] |> List.sort (fun (_, freq1) (_, freq2) -> freq2 - freq1) in
    let selected_ngrams = List.take (dictionary_size / 2) sorted_frequencies |> List.map fst in
    let intervals =
      List.sort String.compare selected_ngrams
      |> List.fold_left
           (fun acc ngram ->
             match acc with
             | [] -> [ngram, None]
             | (last_ngram, _) :: rest ->
                 let start = last_ngram ^ String.make 1 (char_of_int (int_of_char (String.get last_ngram (String.length last_ngram - 1)) + 1)) in
                 let end_ = ngram in
                 (start, Some end_) :: (ngram, None) :: rest)
           []
    in
    intervals
end